### ISSUE TITLE
DataContractSerializer uses Reflection.Emit

### ISSUE THREAD

**Issue by MichalStrehovsky** [dotnet/runtimelab#195]
We need the build configuration that .NET Native used to use.

JIT/Performance/CodeQuality/Serialization/Deserialize
JIT/Performance/CodeQuality/Serialization/Serialize

**Comment by queequac** [comment]
Stumbled over exactly this issue.
While I was used to one of the packages I depend on working fine with UWP, it took me quite a while to find the root cause being the DataContractSerializer used inside in that package. Unfortunately the library was eating the `System.PlatformNotSupportedException Dynamic code generation is not supported on this platform.` exception.

Would appreciate a fix for this, because aside of this NativeAOT works like a charm. 👍

**Comment by MichalStrehovsky** [comment]
dotnet/runtime#41759 is where the build configuration for UWP was deleted. So the task at hand is to figure out the changes to the System.Private.DatacontractSerialization project (I think) that we need to make the library build as it used to for UWP.

I guess we need to add/remove a couple `#define`s. It's possible the defines got deleted in the meantime too, so we might need to restore that code and clean up any bit-rot.

**Comment by mconnew** [comment]
There is an internal property to enable an alternative serializer which doesn't need RefEmit. We had an #if to make it public on UWP and emitted something on UWP which flipped it to use the reflection based serializer instead of the RefEmit version. That way it would just work and you don't get the exception.

**Comment by MichalStrehovsky** [comment]
Thank you! I think we'll want to key this decision off from [`RuntimeFeature.IsDynamicCodeSupported`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.runtimefeature.isdynamiccodesupported?view=net-5.0) eventually and have IL Linker trim unused code based on the value of this for scenarios where size matters. Those pieces are already in place.

**Comment by queequac** [comment]
Any update on this one? Seems that the latest dotnet6 bits are still based on reflection.emit only.
Would be eager to try it out once the serializer is ready for aot again. But I have no clue when something could be expected. :)

**Comment by mconnew** [comment]
@StephenMolloy for visibility as he's the owner of the serializers.

**Comment by MichalStrehovsky** [comment]
Fixed in #684. There were a lot of `uapaot` ifdefs in the DataContractSerializer codebase that were deleted, but based on a cursory look they all looked related to pregenerated serialization code (which is not applicable). NuGet package with the fix should be out within 24 hours or so. Shout if it's still broken in the new package.

**Comment by queequac** [comment]
Good news: `PlatformNotSupportedException `is gone.
Bad news: Fails with the following exception.😉

`
MakeGenericMethod() cannot create this generic method instantiation because no code was generated for it: 'System.Runtime.Serialization.FastInvokerBuilder.CreateSetterInternal<Opc.Ua.ApplicationConfiguration,Opc.Ua.ApplicationType>(System.Reflection.PropertyInfo)'.
`

Is that something I could cure with properly configuring rd.xml?

**Comment by MichalStrehovsky** [comment]
> Is that something I could cure with properly configuring rd.xml?

Yes, you could add a `Method` entry for the method in question with the given generic arguments (https://github.com/dotnet/runtimelab/blob/feature/NativeAOT/docs/using-nativeaot/rd-xml-format.md). Depending on how many valuetype properties you have, it might be rather annoying to deal with. I've submitted #692 that should take care of this problem in a more general way.

**Comment by queequac** [comment]
Thanks, your #692 made life indeed easier.👍

Only interesting finding, even though DataMember is set to optional (e.g. `[DataMember(IsRequired = false, EmitDefaultValue = false, Order = 4)]`), it fails at runtime with `Error in line 5 position 22. 'Element' 'TraceConfiguration' from namespace 'http://...' is not expected. Expecting element 'ClientConfiguration'.`. So I have to add the defaults to my XML at the moment to fill the missing gaps, making the serializer happy when compiled with AOT.

**Comment by jkotas** [comment]
@queequac Could you please create a small repro that demonstrates the problem and open a new issue on it?
